tests/contrib/openstack/test_os_contexts.py

```diff
@@ -4395,6 +4395,7 @@ def __init__(self, address):
 TEST_CPULIST_1 = "0-3"
 TEST_CPULIST_2 = "0-7,16-23"
 TEST_CPULIST_3 = "0,4,8,12,16,20,24"
+TEST_CPULIST_4 = "0-7,^5,16-23,^20,^22"
```

Comment:
Would it also make sense to add some not well formed strings and test if the code behaves as you would expect?

Reply:
I can but I know it will not behave as expected ;-). I re-used _parse_cpu_list(cpulist) as is but I can make it more robust if needed.

Reply:
> I can but I know it will not behave as expected ;-). I re-used _parse_cpu_list(cpulist) as is but I can make it more robust if needed.

The concepts of "assume all data is corrupt" and thus "parse and reject malformed data" come to mind here.  Fuzzy testing tends to break code for a reason :)  I agree with @fnordahl that some invalid strings should be used to test the effectiveness of parsing the data out of the string.

Reply:
No problem, will do!

Reply:
Tests added with invalid data.